[12.x] Improve Queues and Horizon documentation (#10596)
laravel/docs: horizon.md

When you start Horizon, it will use the worker process configuration options for the environment that your application is running on. Typically, the environment is determined by the value of the `APP_ENV` [environment variable](/docs/{{version}}/configuration#determining-the-current-environment). For example, the default `local` Horizon environment is configured to start three worker processes and automatically balance the number of worker processes assigned to each queue. The default `production` environment is configured to start a maximum of 10 worker processes and automatically balance the number of worker processes assigned to each queue.

You may also define Horizon's environment independently of the application's `APP_ENV` by adding an `env` key to your Horizon configuration file:

```php

return [
    /*
    | ------------------------------------------------------------------

>     | Horizon Environement

Reviewer: spelling: "Environment"

>     | Horizon Environement

Reviewer: spelling:

```suggestion
    | Horizon Environment
```

Within Horizon's default configuration file, you will notice a `defaults` configuration option. This configuration option specifies the default values for your application's [supervisors](#supervisors). The supervisor's default configuration values will be merged into the supervisor's configuration for each environment, allowing you to avoid unnecessary repetition when defining your supervisors.

<a name="queue-priorities"></a>
### Queue Priorities

Each supervisor can process one or more queues. Just like Laravel's default queue system, queues are processed in the order they are listed in the queue configuration option, giving higher priority to queues defined first:

```php
'environments' => [
    'production' => [
        'supervisor-1' => [
            'connection' => 'redis',
            'queue' => ['high', 'default', 'low'],
        ],
    ],
],
```

> In this example, jobs in the `high` queue will be prioritized over those in the `default` or `low` queues.

Reviewer 2: I think this is only the case if `balance=false`.

Author: You are right, I need rewrite this section

Author: I've just completely rewritten the section about the strategies.

You can control Horizon's ability to scale worker processes using the `minProcesses` and `maxProcesses` options:

- `minProcesses` defines the minimum number of worker processes in total. This value must be greater than or equal to 1.
- `maxProcesses` defines the maximum total number of worker processes Horizon may scale up to.

<a name="max-job-attempts"></a>
### Max Job Attempts

> [!NOTE]
> Before refining these options, make sure you are familiar with Laravel's base [queue services](/docs/{{version}}/queues#max-job-attempts-and-timeout) and the concept of 'attempts'.

You can define the maximum number of attempts a job can consume directly at the supervisor level:

```php
'environments' => [
    'production' => [
        'supervisor-1' => [
            // ...
            'tries' => 10,
        ],
    ],
],
```

> [!NOTE]
> This option is similar to the `--tries` option when using the Artisan command to process queues.

Fine-tuning the `tries` option is essential when using middlewares such as `WithoutOverlapping` or `RateLimited` because they consume attempts. However, depending on your case, it may be more appropriate to set the job class' `$tries` property.

> 

Reviewer 3: ```suggestion

If `$tries` is defined on the job class, then it will take precedence over the `tries` provided in the Horizon configuration.

```

Horizon uses the framework `WorkCommand`:

https://github.com/laravel/horizon/blob/2744ca84896012d2de135ea0cbb50a8da26d3827/src/Console/WorkCommand.php#L7

Which passes in `WorkerOptions`:

https://github.com/laravel/framework/blob/c82b43b757c608c0afc1b549aa802b12cc847f7c/src/Illuminate/Queue/Worker.php#L425

And only uses the `WorkerOptions::$maxTries` if `null` is returned from the `maxTries()` method on the job (which returns its `$tries` property if its set):

https://github.com/laravel/framework/blob/c82b43b757c608c0afc1b549aa802b12cc847f7c/src/Illuminate/Queue/Worker.php#L433-L435

https://github.com/laravel/framework/blob/c82b43b757c608c0afc1b549aa802b12cc847f7c/src/Illuminate/Queue/Worker.php#L516-L518

Author: I tried to work this into the other sentences to keep it concise.

If you do not specify a value for the `tries` option, jobs will only be attempted once or as many times as specified by the job class `$tries` property.

> 

Reviewer 3: ```suggestion

If `tries` (or `$tries` on the job class) is set to 0, the job may be attempted (and released) an unlimited number of times. This can be helpful in handling circumstances when the number attempts may be unknown.
```

Not sure if we want to recommend setting `$maxExceptions` here to support setting tries to 0? Thoughts?

Author: You're right,  I'll try to reformulate this section to add context about `$maxExceptions` and `tries=0` while keeping it concise

- `maxProcesses` defines the maximum total number of worker processes Horizon may scale up to.

<a name="max-job-attempts"></a>
### Max Job Attempts

> [!NOTE]
> Before refining these options, make sure you are familiar with Laravel's base [queue services](/docs/{{version}}/queues#max-job-attempts-and-timeout) and the concept of 'attempts'.

You can define the maximum number of attempts a job can consume directly at the supervisor level:

```php
'environments' => [
    'production' => [
        'supervisor-1' => [
            // ...
            'tries' => 10,
        ],
    ],
],
```

> [!NOTE]
> This option is similar to the `--tries` option when using the Artisan command to process queues.

Fine-tuning the `tries` option is essential when using middlewares such as `WithoutOverlapping` or `RateLimited` because they consume attempts. However, depending on your case, it may be more appropriate to set the job class' `$tries` property.

If you do not specify a value for the `tries` option, jobs will only be attempted once or as many times as specified by the job class `$tries` property.

<a name="job-timeout"></a>
### Job Timeout

> Similarly, you can set a `timeout` value at the supervisor level, which specifies how many seconds a worker process can run a job before it’s forcefully terminated. Once terminated, the job will either be retried or marked as failed, depending on your queue configuration.

Reviewer: looks like you've got a special apostrophe character in there:

```suggestion
Similarly, you can set a `timeout` value at the supervisor level, which specifies how many seconds a worker process can run a job before it's forcefully terminated. Once terminated, the job will either be retried or marked as failed, depending on your queue configuration.
```

The `auto` strategy, which is the configuration file's default, adjusts the number of worker processes per queue based on the current workload of the queue. For example, if your `notifications` queue has 1,000 pending jobs while your `render` queue is empty, Horizon will allocate more workers to your `notifications` queue until the queue is empty.

> When using the auto scaling strategy, you may define the `minProcesses` and `maxProcesses` configuration options:

Reviewer 2: In my testing these work with any balacing mode or at least also with `balance=false`.

Author: Yes, that's true, but these options do not have the same meaning. To make more sense of it, I've added the `processes` key, which is an alias for `maxProcesses`

Author: @peaklabs-dev  Ok, I just realized that I didn't understand the `false` option correctly. So I guess it was really confusing.

Really appreciate your reviews here! I've updated section about this strategy, let me know if it's sounds good to you.

PS: I’m pretty sure I don’t fully understand the `simple` one either. From what you’re saying, I’m guessing that Horizon always assigns the same number of processes to each queue when scaling up?

Author: Did some testing and it seems that when using the simple strategy, there is no scaling. All worker processes defined by maxProcesses are always running. Let me know if I'm wrong
